### PR TITLE
2.0 alpha

### PR DESCRIPTION
**Pull request by eldh**
Ideas for a 2.0 alpha. Actions and logging are now using "official" plugin hooks. Updated implementation, tests and example app. More hooks for plugins might be needed, but works for the ones we have now at least.

* Major internal re-write to improve modularity and enable extensions.
* Breaking: `Base` is renamed to `Tide`
* Breaking: Changed the API for adding actions: `Tide.addActions` is remove in favor of `initActions`
* Breaking: Changed api for logging. Logging is now separated into different and need to be enabled one by one. (TODO: Component logging)
* New feature: `Tide.addMiddleWare`, a hook to add middleware that intercepts state changes.
* New feature: `Tide.addProp`, a hook to safely monkey patch props onto a tide instance.
* New feature: `Tide.addComponentProp`, a hook to add properties to the props of `TideComponent`.
* New feature: `NamespacedActions`, an alternative actions plugin that lets you define namespaces for your actions classes.

Feedback much appreciated ofc.

### CODE REVIEW

File: src/namespacedActions.js [hunk 1]
```diff
@@ -0,0 +1,64 @@
+import Actions from './actions'
+
+export default class NamespacedActions extends Actions {
```
Comment (JoelBesada) [inline src/namespacedActions.js]: What's the reasoning for not just making this the default `Actions`? The behaviour stays the same as before if you don't set a namespace, so I can't tell what the benefits are of having this as a separate class.

Comment (eldh) [inline src/namespacedActions.js]: Yeah, I thought about that and it's probably a good idea. I wanted to think a bit more about possible changes to the the api for that though, but I think it's something we should do before final release.

File: src/base.js [hunk 2]
```diff
@@ -1,30 +1,29 @@
-import assign from 'lodash.assign'
-import diff from 'immutablediff'
-import forEach from 'lodash.foreach'
-import isArray from 'lodash.isarray'
-import keys from 'lodash.keys'
-import wrap from 'lodash.wrap'
 import defer from 'lodash.defer'
```
Comment (JoelBesada) [inline src/base.js]: Why don't we remove this dependency as well?

Comment (eldh) [inline src/base.js]: Not sure if there's a good alternative? Perhaps rolling our own with setTimeout is good enough in this case?

Comment (JoelBesada) [inline src/base.js]: Yeah it's basically just a `setTimeout(cb, 1)` call https://github.com/lodash/lodash/blob/master/defer.js

Comment (eldh) [inline src/base.js]: Ah, simple enough

**Comment by eldh** [comment]
Replaced old actions with `NamespacedActions` and removed `lodash.defer` dependency.

**Comment by eldh** [comment]
wdyt @JoelBesada ? good enough to release an alpha version?

**Comment by JoelBesada** [comment]
Yes, let's do it!

**Pull request merged**